### PR TITLE
Add option `OnlyWhenUsingAnyOf` to the `Bundler/GemComment` cop.

### PR DESCRIPTION
**Pull request by ric2b**
This PR adds a boolean option `OnlyWhenUsingAnyOf` to the `Bundler/GemComment` cop. 

The goal of this option is to require gem comments only if the gem version has an explicit version specifier or other options like `required/github/etc`, so that the reason for the specifier/options is documented.

I tested it on a project I work on and it worked as expected.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by ric2b** [comment]
By the way, I'm also planning on doing a similar PR for parameters that change the source of a gem, such as `source:`, `github:`, etc ([documented here](https://bundler.io/v2.0/man/gemfile.5.html#SOURCE)).

I was thinking it could be a bit more general and work by receiving a list of parameter names that should activate the cop, something like: `OnlyIfUsingOptions: ['require', 'github']

Does that make sense to you?

File: lib/rubocop/cop/bundler/gem_comment.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ module Cop
     module Bundler
       # Add a comment describing each gem in your Gemfile.
       #
-      # @example
+      # @example OnlyIfVersionRestricted: false (default)
```
Comment (bbatsov) [inline lib/rubocop/cop/bundler/gem_comment.rb:8]: I'd also expand the cop summary.

**Comment by bbatsov** [comment]
You should also mention this in the default config. I'm not sure I like the idea of a boolean option, though - might be a better idea to have some styles like "always", "only gems with some extra params". I like your idea overall, but we have to be careful to make the config both easy to understand and extendable.

**Comment by ric2b** [comment]
Yeah, making it extensible is a great point.

What do you think of a configuration with three parameters: 
- AllGems: Boolean
- WithVersionSpecifiers: Boolean
- WithAnyOfTheseOptions: Array

Where `AllGems: true` would ignore the other options and apply to all gems. Otherwise the rest of the options would be combined.

As an example:
```
Bundler/GemComment:
  Enabled: true
  AllGems: false
  WithVersionSpecifiers: true
  WithAnyOfTheseOptions: ['github', 'bitbucket']
```

Would only trigger for a gem that has version restrictions OR uses at least one of `github:` and `bitcucket:`.

**Comment by bbatsov** [comment]
The problem with the boolean flags approach is that some of them are not compatible with one another and this makes the configuration harder. A different approach might be to just have a list of things to check - e.g.:

```
- Check
  - gems_with_version
  - gems_from_source
  - etc
```

And I guess when empty this can mean "all". Or there can be an enforced style "All" and "Subset" and the second one would consult this extra configuration option. Not sure what's the best approach, but I really don't like the boolean flags, so I'd rather us avoid it.

**Comment by ric2b** [comment]
My last suggestion would make the options compatible with each other, it's just that `gems_with_version` would be a separate option because it's not a kwarg. 

But I like your suggestion, it makes the configuration more straightforward, just need to treat version specifiers as a special case 👍

**Comment by bbatsov** [comment]
Fine by me.

**Comment by ric2b** [comment]
@bbatsov The new version has a single new configuration: `OnlyWhenUsingAnyOf`. 

If it's empty the cop behaves the same as before, checking all gems, otherwise it only registers offenses on gems with one or more of the configured options. 

`with_version_specifiers` can also be passed as an option in to check for version specifiers.

Happy to change the names to whatever you prefer, if you don't think they're clear enough or consistent with rubocop conventions.

File: lib/rubocop/cop/bundler/gem_comment.rb [hunk 2]
```diff
@@ -58,6 +92,28 @@ def ignored_gem?(node)
           ignored_gems = Array(cop_config['IgnoredGems'])
           ignored_gems.include?(node.first_argument.value)
         end
+
+        def checked_options_present?(node)
+          cop_config[CHECKED_OPTIONS_CONFIG].include?(VERSION_SPECIFIERS_OPTION) && version_specified_gem?(node) \
+          or contains_checked_options?(node)
```
Comment (mwolman) [inline lib/rubocop/cop/bundler/gem_comment.rb:98]:
```suggestion
          cop_config[CHECKED_OPTIONS_CONFIG].include?(VERSION_SPECIFIERS_OPTION) &&
            version_specified_gem?(node) ||
            contains_checked_options?(node)
```

File: lib/rubocop/cop/bundler/gem_comment.rb [hunk 3]
```diff
@@ -58,6 +92,28 @@ def ignored_gem?(node)
...
+        end
+
+        # Besides the gem name, all other *positional* arguments to `gem` are version specifiers,
+        # as long as it has one we know there's at least one version specifier.
+        def version_specified_gem?(node)
+          # arguments[0] is the gem name
+          node.arguments[1]&.str_type? == true
```
Comment (mwolman) [inline lib/rubocop/cop/bundler/gem_comment.rb:105]:
```suggestion
          node.arguments[1]&.str_type?
```

File: spec/rubocop/cop/bundler/gem_comment_spec.rb [hunk 4]
```diff
@@ -63,5 +64,76 @@
         GEM
       end
     end
+
+    context 'and the OnlyWhenUsingAnyOf option is set' do
```
Comment (mwolman) [inline spec/rubocop/cop/bundler/gem_comment_spec.rb:68]:
```suggestion
    context 'when the "OnlyWhenUsingAnyOf" option is set' do
```

File: spec/rubocop/cop/bundler/gem_comment_spec.rb [hunk 5]
```diff
@@ -63,5 +64,76 @@
...
+      before { cop_config['OnlyWhenUsingAnyOf'] = checked_options }
+
+      context 'and version specifiers are checked' do
```
Comment (mwolman) [inline spec/rubocop/cop/bundler/gem_comment_spec.rb:71]:
```suggestion
      context 'when the version specifiers are checked' do
```

File: spec/rubocop/cop/bundler/gem_comment_spec.rb [hunk 6]
```diff
@@ -63,5 +64,76 @@
...
+          GEM
+        end
+
+        it 'does not register an offense if an uncommented gem has only unchecked options' do
+          expect_no_offenses(<<-GEM, 'Gemfile')
+            gem 'rubocop', group: development
+          GEM
+        end
+      end
+    end
```
Comment (mwolman) [inline spec/rubocop/cop/bundler/gem_comment_spec.rb:137]: What do you think about using more contexts in your tests?
https://rspec.rubystyle.guide/#use-contexts

Comment (mwolman) [reply spec/rubocop/cop/bundler/gem_comment_spec.rb:137]: For example this:

```ruby
        it 'registers an offense if an uncommented gem has one of the checked options' do
          expect_offense(<<-GEM, 'Gemfile')
            gem 'rubocop', github: 'some_user/some_fork'
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Missing gem description comment.
          GEM
        end
```

should be something like this:

```ruby
      context 'when an uncommented gem has one of the checked options' do
        it 'registers an offense' do
          expect_offense(<<-GEM, 'Gemfile')
            gem 'rubocop', github: 'some_user/some_fork'
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Missing gem description comment.
          GEM
        end
      end
```

File: lib/rubocop/cop/bundler/gem_comment.rb (hunk 2 above)
Comment (ric2b) [reply lib/rubocop/cop/bundler/gem_comment.rb:98]: I added parenthesis to make it more explicit, I agree that the use of `or` might be confusing with the different order of precedence.

File: spec/rubocop/cop/bundler/gem_comment_spec.rb (hunk 6 above)
Comment (ric2b) [reply spec/rubocop/cop/bundler/gem_comment_spec.rb:137]: Yeah, that looks much better :)

**Comment by ric2b** [comment]
All the tests pass on my machine, the failing test seems completely unrelated to this PR, not sure what to do.

File: config/default.yml [hunk 7]
```diff
@@ -159,6 +159,7 @@ Bundler/GemComment:
     - '**/Gemfile'
     - '**/gems.rb'
   IgnoredGems: []
+  OnlyWhenUsingAnyOf: []
```
Comment (bbatsov) [inline config/default.yml:162]: I think this can be shortened to `OnlyFor`.

File: lib/rubocop/cop/bundler/gem_comment.rb [hunk 8]
```diff
@@ -15,17 +21,45 @@ module Bundler
       #   # Helpers for the foo things.
       #   gem 'foo'
       #
+      # @example OnlyWhenUsingAnyOf: ['with_version_specifiers']
```
Comment (bbatsov) [inline lib/rubocop/cop/bundler/gem_comment.rb:24]: I'd name this just `version_specifiers`.

**Comment by bbatsov** [comment]
I think it's a good idea to summarize in docs somewhere all the common params (https://bundler.io/man/gemfile.5.html) for its easier for people to figure out what's available to them.

**Comment by bbatsov** [comment]
You should also update `VersionChanged` to 0.85.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**